apps/aecore/src/aec_chain_state.erl

```diff
@@ -420,22 +420,21 @@ hash_is_in_main_chain(Hash, TopHash) ->
 %%%-------------------------------------------------------------------
 
 internal_insert(Node, Block, Origin) ->
+    Fun = fun() -> internal_insert_transaction(Node, Block, Origin) end,
+    try
+        aec_db:ensure_transaction(Fun)
+    catch
+        exit:{aborted, {throw, ?internal_error(What)}} -> internal_error(What)
+    end.
+
+internal_insert_transaction(Node, Block, Origin) ->
     case db_find_node(hash(Node)) of
-        error ->
-            %% To preserve the invariants of the chain,
-            %% Only add the block if we can do the whole
-            %% transitive operation (i.e., calculate all the state
-            %% trees, and update the pointers)
-            Fun = fun() -> internal_insert_transaction(Node, Block, Origin)
-                  end,
-            try aec_db:ensure_transaction(Fun)
-            catch exit:{aborted, {throw, ?internal_error(What)}} -> internal_error(What)
-            end;
+        error      -> internal_insert_transaction1(Node, Block, Origin);
         {ok, Node} -> ok;
-        {ok, Old} -> internal_error({same_key_different_content, Node, Old})
+        {ok, Old}  -> internal_error({same_key_different_content, Node, Old})
     end.
 
-internal_insert_transaction(Node, Block, Origin) ->
+internal_insert_transaction1(Node, Block, Origin) ->
```

Comment:
Either this or the calling function should be renamed to capture the true semantics. Just adding a number suffix is not a good idea to get rid of nesting.